README.md
# NAME

Mojolicious::Plugin::Form - abstract forms for Mojolicious and DBIx::Class

<div>
    <a href="https://travis-ci.org/wollmers/Mojolicious-Plugin-Form"><img src="https://travis-ci.org/wollmers/Mojolicious-Plugin-Form.png" alt="Mojolicious-Plugin-Form"></a>
    <a href='https://coveralls.io/r/wollmers/Mojolicious-Plugin-Form?branch=master'><img src='https://coveralls.io/repos/wollmers/Mojolicious-Plugin-Form/badge.png?branch=master' alt='Coverage Status' /></a>
    <a href='http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-Form'><img src='http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-Form.png' alt='Kwalitee Score' /></a>
    <a href="http://badge.fury.io/pl/Mojolicious-Plugin-Form"><img src="https://badge.fury.io/pl/Mojolicious-Plugin-Form.svg" alt="CPAN version" height="18"></a>
</div>

# SYNOPSIS

    $app->plugin('Mojolicious::Plugin::Form');

# DESCRIPTION

[Mojolicious::Plugin::Form](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3AForm) is a Mojolicious-Plugin.

# SEE ALSO

- [Mojolicious](https://metacpan.org/pod/Mojolicious)

# SOURCE REPOSITORY

[http://github.com/wollmers/Mojolicious-Plugin-Form](http://github.com/wollmers/Mojolicious-Plugin-Form)

# AUTHOR

Helmut Wollmersdorfer, <helmut@wollmersdorfer.at>

<div>
    <a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>
</div>

# COPYRIGHT AND LICENSE

Copyright (C) 2013-2020 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Mojolicious-Plugin-Form.

{{$NEXT}}

0.009  2020-08-07 06:00:31 CEST
       - increase minimal version of Mojolicious to 6.10

0.008  2015-04-29 16:22:50 CEST
       - improve tests

0.007  2015-04-28 17:00:16 CEST
       - improve coverage

0.006  2015-04-07 16:30:03 CEST
       - add badges in pod

0.005  2014-08-16 23:39:44 CEST
       - badges in pod

0.004  2014-08-16 09:19:35 CEST
       - adjusted travis

0.003  2014-05-08 14:02:27 CEST
       - release to cpan

0.002  2014-05-03 03:24:22 CEST
       - initial version
       
0.001  2014-05-03
       - Initial release.


t/lib/MyApp.pm
package MyApp;
use Mojo::Base 'Mojolicious';

has db => sub {
  my $self         = shift;

  my $schema = MySchema->new(
    'entries' => {
      'Foo' => MySchema::Class->new(table => 'foo'),
      'Bar' => MySchema::Class->new(table => 'bar'),
    },
  );
  return $schema;
};

sub startup {
  my $app = shift;

  $app->plugin('Mojolicious::Plugin::Form');
  
  $app->helper(schema => sub { shift->app->db });
  
  my $routes = $app->routes;
  $routes->any('/' => sub { shift->render(text => 'Hello World.') } );
  #$routes->any('/dbinfo' => sub { shift->render } );
}

package MySchema;
use Mojo::Base -base;

has 'entries';

sub sources {
  my $self = shift;
  my @sources = (keys %{$self->entries});
  #return (keys %{$self->entries});
  return @sources;
};

sub class {
  my $self = shift;
  my $source = shift;
  return $self->entries->{$source};
};

package MySchema::Class;
use Mojo::Base -base;

has 'table';


package MyApp::Form;
use Mojo::Base 'Mojolicious::Plugin::Form::Base';


package MyApp::Form::Foo;
use Mojo::Base 'MyApp::Form';


package MyApp::Form::Bar;
use Mojo::Base 'MyApp::Form';


package MyApp::Form::Defect;
use Mojo::Base 'MyApp::Form';


1;


lib/Mojolicious/Plugin/Form/Base.pm
package Mojolicious::Plugin::Form::Base;
use Mojo::Base -base;

use Data::Dumper;

has 'elements';
has 'ordered_elements';

has 'element_info' => sub {
  my ($self, $element) = @_;
  return $self->elements->{$element};
};

has 'id_field';

has 'name_field';

has 'order_by';

has 'form_debug' => 0;

sub add_elements {
  my ($self, @elems) = @_;

  my @added;
  my $elements = $self->elements;
  while (my $el = shift @elems) {
    my $element_info = {};
    if ($el =~ s/^\+//) {
      $element_info = $self->element_info($el);
    }

    # If next entry is { ... } use that for the info, if not
    # use an empty hashref
    if (ref $elems[0]) {
      my $new_info = shift(@elems);
      %$element_info = (%$element_info, %$new_info);
    }
    push(@added, $el) unless exists $elements->{$el};
    $elements->{$el} = $element_info;
  }
  push @{$self->ordered_elements}, @added;
  return $self;
}

sub from_schema {
  my $self   = shift;
  my $schema = shift;
  my $source = shift || $self->name;

  $self->order_by || $self->order_by($self->order_field($schema, $source));

  my $columns = [$schema->source($source)->columns];

  #my $result = $schema->resultset($source)->single({$order => $id});

  my $columns_info = $schema->source($source)->columns_info($columns);

  # TODO: smarter
  #my $primary_columns = [$schema->source($source)->primary_columns];
  #$self->id_field($primary_columns->[0]);

  my ($id_field, $name_field) = $self->id_and_name($schema, $source);
  $self->id_field || $self->id_field($id_field);
  $self->name_field || $self->name_field($name_field);

  my $relationships = [ $schema->source($source)->relationships ];

  print STDERR '$relationships: ', Dumper($relationships), "\n" if $self->form_debug;;

  my $rel_elements;
  for my $relation (@$relationships) {
    my $relationship = $self->related($schema, $source, $relation);
    $rel_elements->{$relation} = $relationship if $relationship;
  }

  my $rel_multi;
  for my $relation (@$relationships) {
    my $relationship = $self->multi_related($schema, $source, $relation);
    $rel_multi->{$relation} = $relationship if $relationship;
  }

  $self->elements         || $self->elements({});
  $self->ordered_elements || $self->ordered_elements([]);

  for my $column (@$columns) {
    my $element;

    if (my $rel_element = $self->exchanges_field($rel_elements, $column)) {
      $element = $rel_elements->{$rel_element};
    }
    else {
      my $type = $self->type($columns_info->{$column}->{data_type});
      my $required = $columns_info->{$column}->{is_nullable} ? '' : 'required';

      $element = {
        'type'     => $type,
        'name'     => $column,
        'required' => $required,
      };
      if (exists $columns_info->{$column}->{default_value} && $columns_info->{$column}->{default_value}) {
          $element->{'default'} = $columns_info->{$column}->{default_value};
      }
    }

    if ($column eq $self->id_field) { $element->{'hidden'} = 1; }

    $self->elements->{$column} = $element;
    push @{$self->ordered_elements}, $column;
  }

  for my $multi (keys %$rel_multi) {
    $self->elements->{$multi} = $rel_multi->{$multi};
    push @{$self->ordered_elements}, $multi;
  }

  print STDERR 'id_field: ',$self->id_field,' name_field: ',$self->name_field,"\n" if $self->form_debug;
  print STDERR 'elements: ',Dumper($self->elements),"\n" if $self->form_debug;
  return $self;
}

sub exchanges_field {
  my ($self, $rel_elements, $column) = @_;
  for my $rel_element (keys %$rel_elements) {
    if (exists $rel_elements->{$rel_element}->{'exchanges_field'}
      && $rel_elements->{$rel_element}->{'exchanges_field'} eq $column)
    {
      return $rel_element;
    }
  }
  return;
}

sub related {
  my ($self, $schema, $source, $relation) = @_;
  return undef unless $source;

  #'cond' => { 'foreign.group_id' => 'self.global_role_id' },

  my $rel_info = $schema->source($source)->relationship_info($relation);

  # print STDERR '$rel_info: ', Dumper($rel_info), "\n";

  # TODO: accessor 'multi' (???)
  return undef unless ($rel_info->{attrs}->{accessor} eq 'single');

  my $rel_source =
    $schema->source($source)->related_source($relation)->{'source_name'};

  my @rel_fields = $self->id_and_name($schema, $rel_source);

  my @conditions = %{$rel_info->{cond}};

  my @self_fields = map { /(\w+)$/; $1 } grep {/^(self\.|)(\w+)$/} @conditions;
  my @foreign_fields =
    map { /(\w+)$/; $1 } grep {/^(foreign\.|)(\w+)$/} @conditions;

  my $field_to_exchange = $self_fields[0];
  my $rel_key           = $foreign_fields[0];

  my ($name_field) = grep { $_ !~ /^$rel_key$/ } @rel_fields;

  my $related_element = {
    'type'            => 'Block',
    'nested_name'     => $relation,
    'exchanges_field' => $field_to_exchange,
    'key'             => $rel_key,
    'elements'        => [
      {
        'type' => 'Text',
        'name' => $name_field,
      },
    ],
  };
  return $related_element;
}

sub multi_related {
  my ($self, $schema, $source, $relation) = @_;
  return undef unless $source;

  #'cond' => { 'foreign.group_id' => 'self.global_role_id' },

  my $rel_info = $schema->source($source)->relationship_info($relation);

  # print STDERR '$rel_info: ', Dumper($rel_info), "\n";

  # TODO: accessor 'multi' (???)
  return undef unless ($rel_info->{attrs}->{accessor} eq 'multi');

  my $rel_source =
    $schema->source($source)->related_source($relation)->{'source_name'};

  my @rel_fields = $self->id_and_name($schema, $rel_source);

  my @conditions = %{$rel_info->{cond}};

  my @self_fields = map { /(\w+)$/; $1 } grep {/^(self\.|)(\w+)$/} @conditions;
  my @foreign_fields =
    map { /(\w+)$/; $1 } grep {/^(foreign\.|)(\w+)$/} @conditions;

  my $field_to_exchange = $self_fields[0];
  my $rel_key           = $foreign_fields[0];

  my ($name_field) = grep { $_ !~ /^$rel_key$/ } @rel_fields;

  my $related_element = {
    'type'            => 'Multi',
    'nested_name'     => $relation,
    'exchanges_field' => $field_to_exchange,
    'key'             => $rel_key,
    'elements'        => [
      {
        'type' => 'Text',
        'name' => $name_field,
      },
    ],
  };
  return $related_element;
}

sub order_field {
  my ($self, $schema, $source) = @_;
  return undef unless $source;

  my @columns = $schema->source($source)->columns;
  my @source_ids = grep {/name/} @columns;

  my @primary_columns = $schema->source($source)->primary_columns;
  push @source_ids, @primary_columns;

  #my $table_name = $schema->class($source)->table;
  my $table_name = lc $source;
  push @source_ids, grep {/${table_name}_id/} @columns;

  return $source_ids[0] if (scalar @source_ids);
  return $columns[0]    if (scalar @columns);
}

sub id_and_name {
  my ($self, $schema, $source) = @_;
  return undef unless $source;
  my @columns = $schema->source($source)->columns;

  my @source_ids;
  my @primary_columns = $schema->source($source)->primary_columns;
  push @source_ids, $primary_columns[0];
  push @source_ids, grep {/name/} @columns;

  #push @source_ids, $primary_columns[1] unless (scalar @source_ids >= 2);
  push @source_ids, $columns[1] unless (scalar @source_ids >= 2);

  return @source_ids;
}

sub name {
  my $self       = shift;
  my $class_name = ref $self;
  $class_name =~ s/^.*::(\w+)$/$1/;
  return $class_name;
}

sub type {
  my $self = shift;
  my $data_type = shift || 'text';

  my $data2elem = {
    'integer' => 'number',
    'varchar' => 'text',
    'tinyint' => 'checkbox',
    'enum'    => 'enum',
  };
  return $data2elem->{$data_type} ? $data2elem->{$data_type} : 'text';
}

1;


lib/Mojolicious/Plugin/Form.pm
package Mojolicious::Plugin::Form;
use Mojo::Base 'Mojolicious::Plugin';

use Mojo::Loader qw(data_section find_modules load_class);

our $VERSION = '0.009';

sub register {
  my ($self, $app, $conf) = @_;

  $conf ||= {};

  $app->helper(
    form_handler => sub {
      my $self       = shift;
      my $class_name = shift;

      $class_name ||= 'Default';
      my $namespace = ref($self->app) . '::Form::';

      unless ($class_name =~ m/[A-Z]/) {
        my $namespace = ref($self->app) . '::Form::';
        $namespace = '' if $namespace =~ m/^Mojolicious::Lite/;

        $class_name = Mojo::ByteStream->new($class_name)->camelize;
      }

      $class_name = $namespace . $class_name;
      #my $e = Mojo::Loader->new->load($class_name);
      #my $e = Mojo::Loader->load_class($class_name);
      my $e = load_class($class_name);

      Carp::croak qq/Can't load form '$class_name': / . $e->message
        if ref $e;

      Carp::croak qq/Can't find form '$class_name'/ if $e;

      return $class_name->new(%$conf, @_);
    }
  );
}

1;

__END__

=head1 NAME

Mojolicious::Plugin::Form - abstract forms for Mojolicious and DBIx::Class

=begin html

<a href="https://travis-ci.org/wollmers/Mojolicious-Plugin-Form"><img src="https://travis-ci.org/wollmers/Mojolicious-Plugin-Form.png" alt="Mojolicious-Plugin-Form"></a>
<a href='https://coveralls.io/r/wollmers/Mojolicious-Plugin-Form?branch=master'><img src='https://coveralls.io/repos/wollmers/Mojolicious-Plugin-Form/badge.png?branch=master' alt='Coverage Status' /></a>
<a href='http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-Form'><img src='http://cpants.cpanauthors.org/dist/Mojolicious-Plugin-Form.png' alt='Kwalitee Score' /></a>
<a href="http://badge.fury.io/pl/Mojolicious-Plugin-Form"><img src="https://badge.fury.io/pl/Mojolicious-Plugin-Form.svg" alt="CPAN version" height="18"></a>

=end html

=head1 SYNOPSIS

 $app->plugin('Mojolicious::Plugin::Form');


=head1 DESCRIPTION

L<Mojolicious::Plugin::Form> is a Mojolicious-Plugin.


=head1 SEE ALSO

=over

=item *

L<Mojolicious>

=back

=head1 SOURCE REPOSITORY

L<http://github.com/wollmers/Mojolicious-Plugin-Form>

=head1 AUTHOR

Helmut Wollmersdorfer, E<lt>helmut@wollmersdorfer.atE<gt>

=begin html

<a href='http://cpants.cpanauthors.org/author/wollmers'><img src='http://cpants.cpanauthors.org/author/wollmers.png' alt='Kwalitee Score' /></a>

=end html

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2020 by Helmut Wollmersdorfer

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



